### PR TITLE
Fixed an issue which caused $pristine to get set to false on focus and blur

### LINKED ISSUE #102: Field becomes dirty when using dynamic currency-symbol
### Description

We use angular-locale to set the formatting of currencies etc. I just updated from a pre 0.9 version and noticed the field/form is now dirty after loading the form and since we have a dirty tracker that alerts the user when the form is "unsaved" this alert is now persistent.

### Expected outcome

That the field is not becoming dirty from the currency symbol change.

### Actual outcome

Field becomes dirty.

### Live Demo

The below demo is a modified version of the main example. I changed local to Danish so it will load `kr` by default and then in the scope I change it to `USD`. It's very basic to re-create. If you keep your eyes sharp, you will notice the initial `kr` changing to `USD` and also that the form is first `dirty: false` and then becomes `dirty: true`
http://plnkr.co/edit/oMsst0D0igUst6TOZ6TV?p=preview

### Steps to reproduce

See demo (and the simple code changes). Removing `currency-symbol` from the HTML stops the form from becoming dirty but of course it makes it impossible to use a dynamic currency symbol.

### Browsers Affected

- [x] Chrome
- [x] Firefox
- [x] Safari 9
- [ ] Safari 8 (did not test)
- [ ] Safari 7 (did not test)
- [ ] Edge (did not test)
- [ ] IE 11 (did not test)
- [ ] IE 10 (did not test)

### Off topic:

I kinda like the new way of hiding the currency when the user enters the field, but it seems the model value needs to used for that and not just the view value, which then changes the model and that means the field becomes dirty unless you use the default currency symbol. A hacky way may be to $setPristine() right after loading the symbol, which should be safe enough since the user won't be able to enter the field before, but in theory other code could, which would reset the dirty status, and that could raise the opposite problem/situation.

I also don't like much that the field changes to `22.11` when focusing, as I as would expect it to stay `22,11` which is the Danish formatting (and how it worked pre 0.9.x). I do understand that I can type in `22,12`, but it's confusing for the user to see it change from the correct formatting to what would be considered thousands when editing (by people in countries that uses comma for decimals - and dots for thousands).

In any case, thank you for a great directive!

### PR DESCRIPTION
**Pull request by cecilia-sanare**
- Added `npm prune` to reduce the node_modules cache size
- Updated npm packages
- All packages will now be saved with their exact versions
- Fixed currencyPattern in Angular 1.4.x+

Fixes #102

[Live Demo](https://jsbin.com/fukubopoze/edit?html,output)

### CODE REVIEW

File: src/ng-currency.directive.js [hunk 1]
```diff
@@ -14,6 +14,9 @@ export default function ngCurrency($filter, $locale, $timeout) {
       let initialized, hardCap, min, max, currencySymbol, ngRequired;
       let active = true;
       let fraction = 2;
+      const currencyPatterns = $locale.NUMBER_FORMATS.PATTERNS[1];
+      currencyPatterns.negPre = '(\u00a4';
+      currencyPatterns.negSuf = ')';
```
Comment (nick-woodward) [inline src/ng-currency.directive.js]: @aguirrel 
angular/angular.js#12870
Angular 1.4.x+ changed how the currency filter works. 
Rather then negative values becoming `($1.00)` they become `-$1.00`.

Personally I feel like we should follow Angulars lead on this and use `-$1.00`.

Users can always modify the locale pattern to use parenthesis, 
however if we modify it for them then they don't get a choice.

**Comment by houmark** [comment]
I quickly tested with this branch and while I no longer get a dirty state on the form on load, and I don't get a dirty state on the field when entering it, I also don't get a dirty state when changing the value, so maybe pristine is being too aggressive now?

**Comment by nick-woodward** [comment]
@houmark
I also noticed if you check the model in plnkr or jsbin the $pristine value is false, but the ng-pristine class is present.

**Comment by nick-woodward** [comment]
It feels like there's a lot of inconsistency with ngModel. setting pristine to true doesn't always seem to work.
[Example](https://jsbin.com/bowakafitu/)
* Focus and Blur the field multiple times.

Even if I set it to always set it to pristine it doesn't work. The value eventually becomes false.
Dirty however is consistent to set and can be easily fixed.

**Comment by nick-woodward** [comment]
@houmark 
I think I've resolved all of the issues at this point, can you verify on your end?

**Comment by houmark** [comment]
I really appreciate your work here, but I'm sorry to inform you that I cannot get the field to become $dirty once I change a value in a field with ng-currency attached. I no longer get dirty states on load, and on entering, but when I change a value it does not get dirty, which of course is an issue because we cannot warn about unsaved changes :/ 

What's even more weird is, when I try to use your branch in plunker, I get a dirty form on load, which I don't get in our setup, but things are so weird. In our setup the only thing missing is getting the field / form dirty when the value actually changes by using input (I did leave the field to make sure blur events happened).

Plunker here: http://plnkr.co/edit/oMsst0D0igUst6TOZ6TV?p=preview

**Comment by nick-woodward** [comment]
@houmark 
Weird, it seems the ngModel is getting set to dirty, but the form isn't detecting the change.

**Comment by nick-woodward** [comment]
If you check out the live demo you'll see that the field itself toggles between dirty and pristine correctly, however it seems the form isn't respecting the actual state of the model.

Not entirely sure what the best course of action is since we're essentially trying to trick the model into thinking its not dirty/pristine when we internally modify the view value.

**Comment by houmark** [comment]
Yes, it's overall very weird. Now when I tested more, I suddenly get the form ng-dirty (class) but no fields are dirty on load (I did not change anything in our source code). And yes, when I change the field it does becomes dirty but since the form was dirty on load, it stays dirty. I get mixed results from testing basically. Very weird.

**Comment by houmark** [comment]
As I understand the correct way to get the form to update according to fields in the form is by not using $dirty and $pristine, but only use $setDirty() as that then propagates to the form itself but it looks like this entire dirty/pristine handling is buggy in Angular in certain edge situations (maybe from using the API in a "non-correctish" way).

**Comment by nick-woodward** [comment]
This may end up being a partial rewrite of how ngCurrency works.

I was able to get this little [demo](https://jsbin.com/tibajexoli/edit?html,output) working so that it modifies the visible value 
without causing the form or input to become dirty.

**Comment by houmark** [comment]
Yeah, that's the way, but like you say, it's a whole lot of refactoring needed to comply with the existing features. At the same time, #103 could be taken into consideration so you can load the decimal separator either by locale or by a parameter in the directive.

I'm ready to do extensive testing ;) That's the least I can do!

**Comment by nick-woodward** [comment]
@houmark 
Can you check this out when you get the chance?

Based on the fact I was actually able to write tests around the 
form validation I feel pretty confident it should work.

**Comment by houmark** [comment]
I did just now and unfortunately the form still goes dirty after load - the field may stay pristine or go dirty then pristine.

From looking at the form in the loading process it looks to me like it's the moment where it changes the locale from my Danish set locale for my user to the locale the value is for that field. We use a dynamic locale directive to load user settings to display the formatting according to the current locale. Since 0.9+ it shortly loads "kr" in every ng-currency field and then swaps to "USD" which is the locale for that value in that field. Before 0.9 I never noticed "kr" in the load process. 

I'm sorry I cannot provide more precise trackable info, and our entire form is quite complicated with live socket updates and more like that, but the pain is pre 0.9 everything worked so smooth. The only visual / usability improvement in the newer versions (for us) is the removal of the currency symbol when editing the field, but the tradeoff has been "wrong" formatting of the decimal separator when editing and this pristine/dirty situation which then affects our save reminder service.

I'll be offline some hours from now on but will check in here once I'm back online! Thanks again.

**Comment by nick-woodward** [comment]
How are you pulling in ngCurrency and can you verify its getting the latest version from my GitHub repo?

Also I'm unable to reproduce this anymore, can you try and create a small example using this [jsbin](https://jsbin.com/fukubopoze/edit?html,output)?

**Comment by aguirrel** [comment]
Hi @nick-woodward!

This issue is getting deeper and deeper :smile: Justo to add somethig to this issue.  If I add currency symbol to your jsbin example, $viewValue is always one step behind compared to showed at input.

https://jsbin.com/zapako/1/edit?html,js,output

I will try some other approuch not involving setting or getting $pristine, $dirty status.  Just to play a little.

Thanks a lot!

**Comment by nick-woodward** [comment]
@aguirrel 
I'll take a look into that issue to see if I can resolve it.

At the moment my changes don't involve setting or getting $pristine or $dirty in any way.

**Comment by nick-woodward** [comment]
@aguirrel @houmark 
Just updated this to keep the $viewValue in sync and it looks like setting the $viewValue directly doesn't trigger $dirty or $pristine.
https://jsbin.com/xeyakevoyi/edit?html,output

Please use the following url to pull in my branch:
`https://rawgit.com/nick-woodward/ng-currency/82fe4ca1d86e9d9f407b6aa568e6b6525f171287/dist/ng-currency.js`

This should hopefully prevent any caching issues since its tied to a specific commit.

**Comment by houmark** [comment]
@nick-woodward @aguirrel 

Sorry for not getting back earlier. I was out most of the day.

There's some good news and some bad news :) Not updating the model and only playing with the viewValue is definitely the way to go and this was also what I was trying to suggest from the beginning because I had read elsewhere a long time ago that by only modifying the viewValue there should be no problems with this type of stuff where you display a different value that then actual value which is being "submitted" on save.

First of all the whole dirty / pristine situation seems to be solved completely and I also no longer see a flash of the user currency before the field configured currency loads — Nice!.

But — when I change the value in the field, it has a pretty grave issue with the thousands separator being used as decimal separator increasing the value on blur. See here for a small demo from our form which is using this PR: [GIF demo](https://cl.ly/0s0y3Y0w3I1q)

As long as I don't change the value the field stays pristine (and so does the form) so this is really great progress. If we get the old formatting for numbers according to the locale back and respect that, then everything should be perfect!

P. S. I used npm to pull the latest version like this:
`npm i nick-woodward/ng-currency#master --save` and then I cold reload the entire app and make sure to check to the JS in our bundle to verify it's using the correct edition (by comparing to the dist file in my filesystem). So as long as you also push the dist JS there should be no problem.

**Comment by nick-woodward** [comment]
@houmark 
I'll probably just put back the old parsing logic for now since 
it seemed to cover quite a few scenarios that don't have tests.

**Comment by houmark** [comment]
I'm ready to test ;) I'm here for at least 3-4 more hours.

**Comment by houmark** [comment]
And while user testing is not the same as real written tests, it seems like we have a pretty advanced setup that covers most edge situations. And if it all works out I would recommend a version jump to 3.0 to make sure updating users will know to test their setup to make sure there's no regressions.

**Comment by houmark** [comment]
I tested, and seem to consistently get `NaN` when entering an empty field. But it does keep the value no matter if I use `,` or `.` as decimal separator. So a bit of strengthening the parser and then the display of the correct decimal and we are home safe.

@nick-woodward Check Gitter, I wrote you :)

**Comment by houmark** [comment]
GIF with a demo of NaN: https://cl.ly/091Y0h1o1J39

**Pull request merged**